File: 03-data-structures/go/08-set.go
package main

import (
	"fmt"
)

// Set represents a generic set of elements of type T
// T must be comparable so it can be used as a map key
type Set[T comparable] struct {
	m map[T]struct{} // underlying map stores keys with empty struct{} as dummy value
}

// NewSet creates and returns a new Set with initial capacity
func NewSet[T comparable]() *Set[T] {
	return &Set[T]{
		m: make(map[T]struct{}),
	}
}

// -------------------- Insertion --------------------

// Average Time Complexity: O(1)
func (s *Set[T]) Add(key T) {
	s.m[key] = struct{}{}
}

// -------------------- Search --------------------

// Average Time Complexity: O(1)
func (s *Set[T]) Contains(key T) bool {
	_, exists := s.m[key]
	return exists
}

// -------------------- Deletion --------------------

// Average Time Complexity: O(1)
func (s *Set[T]) Remove(key T) {
	delete(s.m, key)
}

// -------------------- Utilities --------------------

// Returns the number of elements in the set
func (s *Set[T]) Size() int {
	return len(s.m)
}

func (s *Set[T]) ToSlice() []T {
	out := make([]T, 0, len(s.m))
	for k := range s.m {
		out = append(out, k)
	}
	return out
}

// Returns string representation of the set
func (s *Set[T]) String() string {
	keys := make([]T, 0, len(s.m))
	for k := range s.m {
		keys = append(keys, k)
	}
	return fmt.Sprintf("%v", keys)
}

File: 03-data-structures/py/08-set.py
class Set:
    def __init__(self):
        # Use built-in dict to store keys with a dummy value
        self._dict = {}
        self._DUMMY = object()  # Shared dummy value to save memory

    def __repr__(self):
        # Represent only keys in curly braces
        keys = ", ".join(repr(k) for k in self._dict.keys())
        return "{" + keys + "}"

    # -------------------- Search --------------------
    # Time Complexity: O(1) average case
    def contains(self, key):
        """
        Check if key is present in the set.
        """
        return key in self._dict

    def __contains__(self, key):
        return self.contains(key)

    # -------------------- Insertion --------------------
    # Time Complexity: O(1) average case
    def add(self, key):
        """
        Add a key to the set.
        """
        self._dict[key] = self._DUMMY

    # -------------------- Deletion --------------------
    # Time Complexity: O(1) average case
    def remove(self, key):
        """
        Remove a key from the set.
        Raises Exception if key not found.
        """
        if key in self._dict:
            del self._dict[key]
        else:
            raise Exception("Key not found")

    # -------------------- Utilities --------------------
    def size(self):
        """
        Return the number of keys in the set.
        """
        return len(self._dict)
